xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -120,6 +120,13 @@ update_parameters(vector<double> &new_parameters, const vector<double> &gradient
     }
 }
 
+static void
+normalize(vector<double>& gradient, int size) {
```

Comment:
The passed value in the only use of this new function so far is of type `std::vector::size_type`, which on a 64 bit platform is typically a 64 bit unsigned type whereas `int` is typically a 32 bit type (and signed).  So the passed value might not fit in an `int` for really large inputs (more than ~2 billion training data items) at which point this function would quietly do the wrong thing.  That's quite an extreme input size, but probably not completely inconceivable (and it becomes more conceivable with time - my current computer has 32 million times more RAM than my first computer did!)

In general it's better to use the appropriate type instead.  Using `int` usually works fine for small scale tests but fails eventually, often in hard to debug ways.  C++ doesn't have a generic `size_type` as such but `size_t` is a reasonable choice.

Reply:
fixed in a fixup commit.